validate/validatespec.go

```diff
@@ -0,0 +1,437 @@
+	"io"
+)
+
+//according the function code decode. the last opcode end is not in code stream.
+func verifyBodyWithSpec(fn *wasm.FunctionSig, body *wasm.FunctionBody, module *wasm.Module) (*mockSpecVM, error) {
+	vm := &mockSpecVM{
+		opdStack:   []wasm.ValueType{},
+		ctrlStack:  []ctrlFrame{},
+		code:       bytes.NewReader(body.Code),
+		origLength: uint32(len(body.Code)),
+		curFunc:    fn,
+	}
+
+	localVariables := []wasm.ValueType{}
+	for _, entry := range fn.ParamTypes {
+		localVariables = append(localVariables, entry)
+	}
+
+	for _, entry := range body.Locals {
+		vars := make([]wasm.ValueType, entry.Count)
+		for i := uint32(0); i < entry.Count; i++ {
+			vars[i] = entry.Type
+		}
+		localVariables = append(localVariables, vars...)
+	}
+
+	var fnsig wasm.ValueType
+	if len(fn.ReturnTypes) == 0 {
+		fnsig = wasm.ValueType(wasm.BlockTypeEmpty)
+	} else if len(fn.ReturnTypes) == 1 {
+		fnsig = fn.ReturnTypes[0]
+	} else {
+		return vm, errors.New("MVP only support single return value")
+	}
+
+	err := vm.pushCtrl(wasm.ValueType(fnsig), wasm.ValueType(fnsig), frameTypeFunction)
+	if err != nil {
+		return vm, err
+	}
+
+	for {
+		op, err := vm.code.ReadByte()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return vm, err
+		}
+		opStruct, err := ops.New(op)
+		if err != nil {
+			return vm, err
+		}
+
+		if !opStruct.Polymorphic {
+			if err := vm.adjustStack(opStruct); err != nil {
+				return vm, err
+			}
+		}
+
+		switch op {
+		case ops.Drop:
+			_, err := vm.popOpd()
+			if err != nil {
+				return vm, err
+			}
+		case ops.Unreachable:
+			err := vm.unreachable()
+			if err != nil {
+				return vm, err
+			}
+		case ops.Block:
+			sig, err := vm.fetchByte()
+			if err != nil {
+				return vm, err
+			}
+			err = vm.pushCtrl(wasm.ValueType(sig), wasm.ValueType(sig), frameTypeBlock)
+			if err != nil {
+				return vm, err
+			}
+		case ops.If:
+			sig, err := vm.fetchByte()
+			if err != nil {
+				return vm, err
+			}
+			/*If is not PolymorphicOp. handle operand stack already.*/
+			err = vm.pushCtrl(wasm.ValueType(sig), wasm.ValueType(sig), frameTypeIf)
+			if err != nil {
+				return vm, err
+			}
+		case ops.Loop:
+			sig, err := vm.fetchByte()
+			if err != nil {
+				return vm, err
+			}
+			err = vm.pushCtrl(wasm.ValueType(wasm.BlockTypeEmpty), wasm.ValueType(sig), frameTypeLoop)
+			if err != nil {
+				return vm, err
+			}
+		case ops.Else:
+			/*here seems different with spec. about the function block frame.*/
+			if vm.ctrlSize() == 1 {
+				return vm, errors.New("End can not match function block frame")
+			}
+
+			_, err = vm.matchElse()
+			if err != nil {
+				return vm, err
+			}
+
+			Type, err := vm.popCtrl()
+			if err != nil {
+				return vm, err
+			}
+			err = vm.pushCtrl(Type, Type, frameTypeElse)
+			if err != nil {
+				return vm, err
+			}
+		case ops.End:
+			/*here seems different with spec. about the function block frame.*/
+			if vm.ctrlSize() == 1 {
```

Comment:
this may be caused by the `end` opcode is removed in function body decoding at : https://github.com/go-interpreter/wagon/blob/master/wasm/section.go#L908-L912?
so the `matchFunction` and `matchEnd` function can be removed.

Reply:
acctually at the start of function verify.  pushed a funtion block frame which the type is funtion block type. so the last end at least 2 other than 1

Reply:
the last end added by compiler haved be moved as I commented at the start of verifyBodyWithSpec